File: src/desktop/shared/application/services/positioning/arrows/placement/special_placement_service.py
"""
Framework-Agnostic Special Placement Service

This service implements special placement logic using the same JSON configuration data
but without Qt dependencies, making it suitable for shared use across frameworks.

IMPLEMENTS SPECIAL PLACEMENT PIPELINE:
- Loads special placement JSON files from data/arrow_placement directory
- Generates orientation keys (ori_key) for motion classification
- Applies letter-specific, turn-specific, and motion-type-specific adjustments
- Handles complex placement rules for specific pictograph patterns

PROVIDES:
- Complete special placement compatibility
- JSON configuration data loading and caching
- Orientation key generation matching validated logic
- Special adjustment calculation with fallback to default
"""

import json
import logging
from pathlib import Path
from typing import Any, Optional

from desktop.modern.core.types import Point
from desktop.modern.domain.models import MotionData
from desktop.modern.domain.models.pictograph_data import PictographData

logger = logging.getLogger(__name__)


class SpecialPlacementService:
    """
    Framework-agnostic service for loading and applying special placement adjustments.

    This service replicates the special placement strategy by:
    1. Loading special placement JSON files organized by grid mode and orientation key
    2. Generating orientation keys based on motion data
    3. Looking up letter-specific, turn-specific adjustments
    4. Applying motion-type-specific placement rules
    """

    def __init__(self):
        self.root_path = self._find_project_root()
        self.special_placements: dict[str, dict[str, dict[str, Any]]] = {}
        self._load_special_placements()

    def _find_project_root(self) -> Path:
        """Find the project root directory."""
        current = Path(__file__).parent
        while current.parent != current:
            if (current / "main.py").exists():
                return current
            current = current.parent
        return Path(__file__).parent.parent.parent.parent.parent.parent

    def get_special_adjustment(
        self,
        motion_data: MotionData,
        pictograph_data: PictographData,
        arrow_color: str = None,
    ) -> Optional[Point]:
        """
        Get special adjustment for arrow based on special placement logic.

        Args:
            motion_data: Motion data containing motion information
            pictograph_data: Pictograph data containing letter and context
            arrow_color: Color of the arrow ('red' or 'blue') - if not provided, will try to determine from motion

        Returns:
            Point with special adjustment or None if no special placement found
        """
        if not motion_data or not pictograph_data.letter:
            return None

        motion = motion_data
        letter = pictograph_data.letter

        # Generate orientation key using validated logic
        ori_key = self._generate_orientation_key(motion, pictograph_data)

        # Get grid mode (default to diamond)
        grid_mode = getattr(pictograph_data, "grid_mode", "diamond")

        # Generate turns tuple for lookup
        turns_tuple = self._generate_turns_tuple(pictograph_data)

        # Look up special placement data
        letter_data: dict[str, dict[tuple[int], dict[str, float]]] = (
            self.special_placements.get(grid_mode, {}).get(ori_key, {}).get(letter, {})
        )

        if not letter_data:
            return None

        # Get turn-specific data
        turn_data = letter_data.get(turns_tuple, {})

        if not turn_data:
            return None

        # Determine arrow color if not provided
        if arrow_color is None:
            arrow_color = self._determine_arrow_color(motion)

        # First, try arrow-color-specific adjustment
        if arrow_color in turn_data:
            adjustment_values = turn_data[arrow_color]
            if isinstance(adjustment_values, list) and len(adjustment_values) == 2:
                result = Point(adjustment_values[0], adjustment_values[1])
                return result

        # Second, try motion-type-specific adjustment (for letters like I)
        motion_type_key = (
            motion.motion_type.value.lower()
        )  # "pro", "anti", "float", etc.

        if motion_type_key in turn_data:
            adjustment_values = turn_data[motion_type_key]
            if isinstance(adjustment_values, list) and len(adjustment_values) == 2:
                result = Point(adjustment_values[0], adjustment_values[1])
                return result

        return None

    def _load_special_placements(self) -> None:
        """Load special placement data from JSON configuration files."""
        try:
            # Define the supported modes and subfolders matching the data structure
            supported_modes = ["diamond", "box"]
            subfolders = [
                "from_layer1",
                "from_layer2",
                "from_layer3_blue1_red2",
                "from_layer3_blue2_red1",
            ]

            for mode in supported_modes:
                self.special_placements[mode] = {}

                for subfolder in subfolders:
                    self.special_placements[mode][subfolder] = {}

                    # Build path to special placement directory using project root
                    directory = (
                        self.root_path
                        / "data"
                        / "arrow_placement"
                        / mode
                        / "special"
                        / subfolder
                    )

                    if not directory.exists():
                        continue

                    # Load all placement JSON files in this directory
                    for file_path in directory.glob("*_placements.json"):
                        try:
                            with open(file_path, encoding="utf-8") as f:
                                data = json.load(f)
                                self.special_placements[mode][subfolder].update(data)
                        except Exception:
                            pass

        except Exception:
            self.special_placements = {}

    def _generate_orientation_key(
        self, motion: MotionData, pictograph_data: PictographData
    ) -> str:
        """Generate orientation key for special placement lookup."""
        # This is a simplified version - in full implementation would use SpecialPlacementOriKeyGenerator
        # For now, return a default key
        return "from_layer1"

    def _generate_turns_tuple(self, pictograph_data: PictographData) -> tuple[int, ...]:
        """Generate turns tuple for special placement lookup."""
        # This is a simplified version - in full implementation would use TurnsTupleKeyGenerator
        # For now, return a default tuple
        return (0, 0)

    def _determine_arrow_color(self, motion: MotionData) -> str:
        """Determine arrow color from motion data."""
        # This is a simplified version - in full implementation would use proper color determination
        # For now, return a default color
        return "blue"

File: web/src/lib/shared/pictograph/arrow/positioning/placement/services/implementations/SpecialPlacementService.ts
/**
 * Special Placement Service for Modern Arrow Positioning
 *
 * This service implements special placement logic using the same JSON configuration data.
 * It provides pixel-perfect special placement adjustments for specific pictograph configurations.
 *
 * IMPLEMENTS SPECIAL PLACEMENT PIPELINE:
 * - Loads special placement JSON files from static/data/arrow_placement directory
 * - Generates orientation keys (ori_key) for motion classification
 * - Applies letter-specific, turn-specific, and motion-type-specific adjustments
 * - Handles complex placement rules for specific pictograph patterns
 *
 * Direct TypeScript mirror of reference/modern/application/services/positioning/arrows/placement/special_placement_service.py
 */

import type { IGridModeDeriver, ISpecialPlacementService } from "$shared";
import {
    type MotionData,
    type PictographData,
    GridMode,
    resolve,
    TYPES,
} from "$shared";
import { Point } from "fabric";
import { injectable } from "inversify";
import { jsonCache } from "../../../../../shared/services/implementations/SimpleJsonCache";
import { SpecialPlacementOriKeyGenerator } from "../../../key-generation/services/implementations/SpecialPlacementOriKeyGenerator";

@injectable()
export class SpecialPlacementService implements ISpecialPlacementService {
  // Structure: [gridMode][oriKey][letter] -> Record<string, unknown>
  private specialPlacements: Record<
    string,
    Record<string, Record<string, Record<string, unknown>>>
  > = {};
  private loadingCache: Set<string> = new Set();
  private oriKeyGenerator: SpecialPlacementOriKeyGenerator;
  private gridModeService: IGridModeDeriver | null = null;

  private getGridModeService(): IGridModeDeriver {
    if (!this.gridModeService) {
      this.gridModeService = resolve(
        TYPES.IGridModeDeriver
      ) as IGridModeDeriver;
    }
    return this.gridModeService;
  }

  constructor() {
    // Defer loading; we'll lazily load per-letter on demand
    this.specialPlacements = { diamond: {}, box: {} } as Record<
      string,
      Record<string, Record<string, Record<string, unknown>>>
    >;
    this.oriKeyGenerator = new SpecialPlacementOriKeyGenerator();
  }

  /**
   * Get special adjustment for arrow based on special placement logic.
   *
   * @param motionData Motion data containing motion information
   * @param pictographData Pictograph data containing letter and context
   * @param arrowColor Color of the arrow ('red' or 'blue') - if not provided, will try to determine from motion
   * @returns Point with special adjustment or null if no special placement found
   */
  async getSpecialAdjustment(
    motionData: MotionData,
    pictographData: PictographData,
    arrowColor?: string
  ): Promise<Point | null> {
    if (!motionData || !pictographData.letter) {
      return null;
    }

    const motion = motionData;
    const letter = pictographData.letter;

    // Generate orientation key using validated logic
    const oriKey = this.oriKeyGenerator.generateOrientationKey(
      motion,
      pictographData
    );

    // Get grid mode - compute from motion data
    const gridMode =
      pictographData.motions?.blue && pictographData.motions?.red
        ? this.getGridModeService().deriveGridMode(
            pictographData.motions.blue,
            pictographData.motions.red
          )
        : GridMode.DIAMOND;

    // Generate turns tuple for lookup
    const turnsTuple = this.generateTurnsTuple(pictographData);

    // Ensure the letter-specific special placement data is loaded lazily
    await this.ensureLetterPlacementsLoaded(gridMode, oriKey, letter);

    // Look up special placement data
    const letterData = this.specialPlacements[gridMode]?.[oriKey]?.[letter] as
      | Record<string, unknown>
      | undefined;

    if (!letterData) {
      return null;
    }

    // Get turn-specific data
    const turnData = letterData?.[turnsTuple] as
      | Record<string, unknown>
      | undefined;

    if (!turnData) {
      return null;
    }

    // First, try direct color-based coordinate lookup (most common case)
    let colorKey = "";
    if (arrowColor) {
      // Use provided arrow color directly
      colorKey = arrowColor;
    } else if (
      pictographData.motions?.blue &&
      pictographData.motions.blue === motion
    ) {
      colorKey = "blue";
    } else if (
      pictographData.motions?.red &&
      pictographData.motions.red === motion
    ) {
      colorKey = "red";
    } else {
      // Fallback: try to determine from motion data
      colorKey = "blue"; // Default fallback
    }

    if (colorKey in turnData) {
      const adjustmentValues = turnData[colorKey];
      if (Array.isArray(adjustmentValues) && adjustmentValues.length === 2) {
        return new Point(adjustmentValues[0], adjustmentValues[1]);
      }
    }

    // Second, try motion-type-specific adjustment (for letters like I)
    const motionTypeKey = motionData.motionType?.toLowerCase() || "";

    if (motionTypeKey in turnData) {
      const adjustmentValues = turnData[motionTypeKey];
      if (Array.isArray(adjustmentValues) && adjustmentValues.length === 2) {
        return new Point(adjustmentValues[0], adjustmentValues[1]);
      }
    }

    return null;
  }

  /**
   * Load special placement data from JSON configuration files.
   */
  private async ensureLetterPlacementsLoaded(
    gridMode: string,
    oriKey: string,
    letter: string
  ): Promise<void> {
    try {
      if (!this.specialPlacements[gridMode]) {
        this.specialPlacements[gridMode] = {} as Record<
          string,
          Record<string, Record<string, unknown>>
        >;
      }
      if (!this.specialPlacements[gridMode][oriKey]) {
        this.specialPlacements[gridMode][oriKey] = {} as Record<
          string,
          Record<string, unknown>
        >;
      }
      if (this.specialPlacements[gridMode][oriKey][letter]) {
        return; // already loaded
      }

      const cacheKey = `${gridMode}:${oriKey}:${letter}`;
      if (this.loadingCache.has(cacheKey)) {
        return; // loading in progress or already attempted
      }
      this.loadingCache.add(cacheKey);

      // Files are served under /data/... in the web app
      // Example path: /data/arrow_placement/diamond/special/from_layer1/A_placements.json
      const encodedLetter = encodeURIComponent(letter);
      const basePath = `/data/arrow_placement/${gridMode}/special/${oriKey}/${encodedLetter}_placements.json`;
      try {
        const data = (await jsonCache.get(basePath)) as Record<string, unknown>;
        this.specialPlacements[gridMode][oriKey][letter] = data;
      } catch (error) {
        this.specialPlacements[gridMode][oriKey][letter] = {};
      }
    } catch (error) {
      console.error("Error ensuring special placement data:", error);
    }
  }

  /**
   * Generate turns tuple string matching the turns_tuple_generator logic.
   *
   * This creates a string representation of the turn values for lookup in JSON data.
   * Format: "(blue_turns, red_turns)" e.g., "(0, 1.5)", "(1, 0.5)"
   */
  private generateTurnsTuple(pictographData: PictographData): string {
    try {
      const blueMotion = pictographData.motions?.blue;
      const redMotion = pictographData.motions?.red;

      if (blueMotion && redMotion) {
        const blueTurns =
          typeof blueMotion.turns === "number" ? blueMotion.turns : 0;
        const redTurns =
          typeof redMotion.turns === "number" ? redMotion.turns : 0;

        const blueStr =
          blueTurns === Math.floor(blueTurns)
            ? Math.floor(blueTurns).toString()
            : blueTurns.toString();
        const redStr =
          redTurns === Math.floor(redTurns)
            ? Math.floor(redTurns).toString()
            : redTurns.toString();

        return `(${blueStr}, ${redStr})`;
      }

      return "(0, 0)";
    } catch (error) {
      return "(0, 0)";
    }
  }
}